Add polygon primitive

mdales opened the pull request:
I was looking to use this for a specific application that needs polygon support, so I've added a basic Polygon shape type. 

I manually tested this under both macOS 13 and Fedora 37, and the example in the snippet works (tested both filled and unfilled).

mdales committed:
Add polygon primative

fwcd commented on Sources/CairoGraphics/context/CairoContext.swift:
```diff
@@ -100,6 +100,26 @@ public final class CairoContext: GraphicsContext {
+
+        context.move(to: polygon.points.first!.asTuple)
+        for point in polygon.points.suffix(polygon.points.count - 1) {
+            context.line(to: point.asTuple)
```
I think, we should close off the path, at least when filling is disabled (or, if this is intended, perhaps rename `Polygon` to something more general, e.g. `Path`). Same goes for the CoreGraphics implementation.

mdales replied:
Good idea. I did the change in the Polygon itself, so the logic was just in one place, and added some tests. I also added a check that it wasn't already closed, which was the case for the geospatial data I'm using this to render.

fwcd reviewed (commented):
Neat, thanks for looking into this. Looking good for the most part, just one remark below

mdales committed:
Close polygon when filled if not already so.

mdales committed:
Always close polygons.

mdales commented:
Whilst on vacation, I thought more about your comment @fwcd, and realised that an open polygon is not really a polygon, and so now close them regardless of whether they're filled or not. If people want a path, then a new shape could be added for that.

mdales committed:
Allow multiple paths on polygon for holes.

Both CoreGraphics and Cairo seem to follow the usual clockwise are
positive and counter-clockwise paths are negative, so I've not tried
to add additional logic here.

At some point it may be desirable to have fill method options to let
the user select between winding and even-odd, but both backends
default to winding count.

fwcd commented on Sources/Graphics/shape/Polygon.swift:
```diff
@@ -0,0 +1,36 @@
+import Utils
+
+public struct Polygon<T: IntExpressibleAlgebraicField> {
+    public let paths: [[Vec2<T>]]
```
Do we want polygons to be able to represent disjoint shapes? Is there any advantage over just creating multiple polygons in that case?

mdales replied:
This is to let you create polygons with holes in (i.e., you have a void in your polygon where the data already in the image shows through). Both CoreGraphics and Cairo let you provide multiple subpaths and when you then call fill, it'll use either the winding rule or the even/odd method to create the holes. 

If you want to overlap or have individual polygons you shouldn't use this method to stack them up, you should just make separate calls to draw(polygon:...) for that.

It's a bit of an odd interface, but both Cairo and CoreGraphics do the same thing here essentially. The only other API for this I've seen in use is where the first path in the list is treated as the outside, and every subsequent path is assumed to be a hole, but that's now how either of the supported backends work, so I didn't go for that.

fwcd replied:
Ah okay, thanks for clarifying, that sounds reasonable!

fwcd commented:
Could you rebase this on top of `main` (and perhaps the other PR too)? That way I could run CI on your branches

mdales commented:
Rebased! I'll do the other one tomorrow as part of working in your feedback 👍

fwcd commented:
Thanks!
